### PR TITLE
5.x spid

### PR DESCRIPTION
**Pull request by thomaschiozzi-tndigit**
Initial version for authentication with SPID SAML identity providers.

### CODE REVIEW

File: src/main/java/it/smartcommunitylab/aac/crypto/CertificateParser.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+
+public class CertificateParser {
+
+    // updatePemHeaderFooter ensured that a PEM (as UTF-8 string) starts
+    // and terminated with the appropriate footer. Which footer is appropriate
+    // depends in the the params kind
+    // TODO: kind should be enum
+    private static String updatePemHeaderFooter(String value, String kind) {
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/crypto/CertificateParser.java:23]: drop the methods manipulating user input. We can try guessing content if needed for parsing, but we won't fix/correct user input

File: src/main/java/it/smartcommunitylab/aac/spid/auth/SpidMetadataResolver.java [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+        boolean isPublic = true;
+
+        ContactPerson contactPerson = contactPersonBuilder.buildObject();
+        contactPerson.setType(ContactPersonTypeEnumeration.OTHER);
+
+        EmailAddress emailAddress = emailAddressBuilder.buildObject();
+        emailAddress.setURI(configMap.getContactPerson_EmailAddress());
+        contactPerson.getEmailAddresses().add(emailAddress);
+
+        //        OPTIONAL
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/spid/auth/SpidMetadataResolver.java:338]: why are OPTIONAL commented? Please try to either add or drop them.

File: src/main/java/it/smartcommunitylab/aac/spid/auth/SpidResponseValidator.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+//import org.opensaml.saml.saml2.core.NameIDType;
+//import org.opensaml.saml.saml2.core.Response;
+//import org.opensaml.saml.saml2.core.Status;
+//import org.opensaml.saml.saml2.core.StatusCode;
+//import org.opensaml.saml.saml2.core.Subject;
+//import org.springframework.security.saml2.core.OpenSamlInitializationService;
+//import org.springframework.security.saml2.provider.service.authentication.OpenSaml4AuthenticationProvider;
+//import org.springframework.util.CollectionUtils;
+//
+//public class SpidResponseValidator {
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/spid/auth/SpidResponseValidator.java:31]: this whole class is commented: if it's not used drop it.

File: src/main/java/it/smartcommunitylab/aac/spid/auth/SpidWebSsoAuthenticationRequestFilter.java [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        this(
+            SystemKeys.AUTHORITY_SPID,
+            providerConfigRepository,
+            relyingPartyRegistrationRepository,
+            DEFAULT_FILTER_URI
+        );
+    }
+
+    public SpidWebSsoAuthenticationRequestFilter(
+        String authorityId,
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/spid/auth/SpidWebSsoAuthenticationRequestFilter.java:99]: you can drop the customizable authorityId from every class, this is SPID specific we do *want* to avoid reuse

File: src/main/java/it/smartcommunitylab/aac/spid/provider/SpidIdentityProviderConfigMap.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    private String contactPerson_Type; // "other" (mandatory), optionally includes "billing" (unless private SP, in which case "billing" is also mandatory)
+
+    // AAC options
+    // NOTE: only one among {idps, idpMetadataUrl} can be non null
+    // NOTE: idps is intended to be a subset of the local SPID registry. Both idps and idpMetadataUrl are intended
+    //  to be used for testing purposes. If both are null, the full local SPID registry will be used instead.
+    private Set<String> idps; // optional (see note above)
+    private String idpMetadataUrl; // optional (see note above)
+    private Set<SpidAttribute> spidAttributes; // optional
+    private SpidAuthnContext authnContext; // mandatory
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/spid/provider/SpidIdentityProviderConfigMap.java:83]: add validation annotations to mandatory fields (NotNull/NotBlank etc). Also add regex validation (Pattern) to mailAddress,phone,ipacode etc

File: src/main/java/it/smartcommunitylab/aac/spid/provider/SpidLoginProvider.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import javax.validation.constraints.NotBlank;
+import org.apache.commons.lang3.ArrayUtils;
+
+public class SpidLoginProvider extends AbstractLoginProvider {
+
+    public static final String[] ICONS = { "google", "facebook", "github", "microsoft", "apple", "instagram" };
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/spid/provider/SpidLoginProvider.java:29]: drop logic for matching on name, spid is SPID and uses the mandated button

File: src/main/java/it/smartcommunitylab/aac/spid/registry/IconStore.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.lang.Nullable;
+
+/*
+ * IconStore is a utility class used to recover the icon associated to a
+ * SPID identity provider. Icon URLs are obtained from a request with a request
+ * with pattern
+ *  https://registry.spid.gov.it/entities-idp/{entityId}?output=json
+ */
+public class IconStore {
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/spid/registry/IconStore.java:41]: this should be refactored:
* sourcing metadata from a remote address is a security risk and should be handled adequately handled
* we should avoid using URLs received from a remote party as img sources, they could link to scripts etc.
* this implementation uses lots of low level or strange classes

File: src/main/java/it/smartcommunitylab/aac/spid/registry/ModelToRegistrationConverter.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public static SpidRegistration Convert(SpidIdpModel model, IconStore icons) {
+        SpidRegistration reg = new SpidRegistration();
+        reg.setMetadataUrl(model.getMetadataUrl());
+        reg.setEntityId(model.getEntityId());
+        reg.setEntityName(model.getEntityName());
+        String iconUrl = icons.GetIcon(model.getEntityId());
+        if (iconUrl != null) {
+            reg.setIconUrl(iconUrl);
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/spid/registry/ModelToRegistrationConverter.java:41]: we can not trust external URLs as sources. If we receive the link from the registry we should dowload, validate and keep a local copy

File: src/main/java/it/smartcommunitylab/aac/spid/registry/RemoteSpidRegistry.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.lang.Nullable;
+
+/*
+ * https://registry.spid.gov.it/assets/data/idp.json
+ */
+public class RemoteSpidRegistry implements SpidRegistry {
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/spid/registry/RemoteSpidRegistry.java:41]: imho you should split all the remote registry part out and do another pull request, all this code can not me merged as-is

File: src/main/resources/public/spid/js/spid-button.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  });
+}
+
+function expandDropdown(provider) {
+  closeOpenItems(provider);
+  let dropdownMenu = document.getElementById("spid-dropdown-" + provider);
+  dropdownMenu.classList.toggle("spid-idp-button-open");
+}
+
+window.onclick = function (event) {
```
Comment (matteo-s) [inline src/main/resources/public/spid/js/spid-button.js:17]: binding an event to the whole window to close a modal is not the best approach. You should really add an overlay layer to the viewport to handle clicking outside.

File: src/main/resources/resources/internal_it.properties [hunk 11]
```diff
@@ -64,6 +64,32 @@ error.registration.wrong_password=Password sbagliata
 error.unauthenticated_user=Utente non autenticato.
 error.unsupported_operation=Operazione non supportata.
 error.wrong_password=Password sbagliata
+error.spid_authentication.2=Autenticazione fallita. Riprova pi� tardi.
```
Comment (matteo-s) [inline src/main/resources/resources/internal_it.properties:67]: the file is coded as Windows1252 not UTF8

File: src/main/java/it/smartcommunitylab/aac/spid/auth/SpidProviderAssertionValidatorBuilder.java [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+    /*
+     * This test is always passed as no expected result should is provided.
+     * See reference:
+     *  https://docs.italia.it/italia/spid/spid-regole-tecniche/it/stabile/metadata.html#service-provider
+     * The SPID test 103 serve the purpose of making the service aware of the
+     * fact that received attributes might be different than requested
+     * attributes and the SP should be written in order to either handle
+     * or accommodate this scenario.
+     */
+    private boolean areRequestedAttributesObtained(
```
[the remark this answers is not in the document]
Comment (thomaschiozzi-tndigit) [reply src/main/java/it/smartcommunitylab/aac/spid/auth/SpidProviderAssertionValidatorBuilder.java:259]: This is reasonable. What should be the expected behaviour if the requested attributes are not obtained? Should we always fail the authentication? Just log the event? Something else?
